doc/proposals/keynames.md

```diff
@@ -0,0 +1,226 @@
+
+## Semi-formal Explanation
+
+### Definitions
+
+- For strings `s` I will use pythons indexing, e.g. `s[0]` is the first character, `s[1]` the second and `s[1:]` is everything but the first character.
+- `c:s` is the string `s` prefixed with the character `c`.
+- `R` is the set of reserved characters.
+- `\` must be part of `R`, but `/` must not be.
+- For all `r` in `R` the following holds:
+  - `I{r}` is a bijection between two sets of strings
+  - If `I{r}(x) = y` then `r = x[0] = y[0]`. `I{r}(x)` is undefined, if `r != x[0]`.
+- `I{\}(e) := e`
+- `I(e)` is defined as `I{e[0]}(e)`, if `e[0]` is in `R`.
+  Otherwise `I(x) = x`.
+  (`I` interprets the escaped part `e` and returns its unescaped form.)
+- `J` is the inverse function of `I`, i.e. `J(u) := e`, if and only if `I(e) = u`.
+- `e` is _e-part-valid_, if and only if `e[0]` is in `R` implies that there is a `u` such that `I(e) = u`.
+- `u` is _u-part-valid_, if and only if `u[0]` is in `R` implies that there is an `e` such that `I(e) = u`.
+
+I ignore the (un)escaping of `/` and `\` here, as it's pretty clear how it works.
+
+### Operations
+
+The following operations can be used to construct a key name from parts and to inspect the last part of the key name.
+This is enough for theoretical reasoning, things like `keySetName` can just be thought of as splitting the argument into parts and using the operations below, even if it should never be implemented like that.
+
+If an operation is "undefined", it means we have an error; an invalid key name.
+
+- `push(k, e, u)`: adds the escaped part `e` and the unescaped part `u` to the key `k`.
+  This cannot be used directly.
+  It is only used to define the other operations.
+- `lastE(k)`: returns the last escaped part of the key `k`.
+- `lastU(k)`: returns the last unescaped part of key `k`.
+- `pushE(k, e)`: adds an escaped part `e` to the key `k`.
+  Defined as:
+  - If `e` is _e-part-valid_ then `pushE(k, e) := push (k, e, I(e))`.
+  - Otherwise `pushE(k, e)` is undefined.
+- `pushU(k, u)`: adds an unescaped part `u` to the key `k`.
+  Defined as:
+  - If `u` is _u-part-valid_ then `pushU(k, u) := push (k, J(u), u)`.
+  - Otherwise `pushU(k, u)` is undefined.
+- `pushL(k, s)`: helper for adding a "literal" part `s` to the key `k`.
+  This can be used to avoid a part having special meaning.
+  Defined as:
+  - If `s[0]` is in `R` then `pushL(k, s) := pushU (k, \:s)`.
+  - Otherwise `pushL(k, s) := push (s, s)`.
+- `putE(k, e)`, `putU(k, u)` and `putL(k, s)` are the same as removing the last part of `k` and then calling `pushE(k, e)`, `pushU(k, u)` and `pushL(k, s)` respectively.
+
+### Storage Plugins
+
+Storage plugins for flat formats (without a native nesting hierarchy) should just use the `keyName` to store values (and metadata, if possible).
+
+Storage plugins with support for a native hierarchy should use the operations `store` and `load` to find the names to use in that hierarchy (e.g. field names in JSON).
```

Comment:
See the note further down in the file.

No, neither the current `keySetBaseName`/`keyBaseName` nor any proposed form have the correct semantics for this. That is why line 70 reads (emphasis added):

> The following operations would be provided in a **separate library** for storage plugins:

Reply:
Yes, but we definitely cannot call it store/load and the API is essential here for this decision (does it integrate nicely to the other API functions).

Do I understand correctly that in essence there would be 3 different forms of key names:

- the "escaped key name" in which storage plugins render key names
- the "unescaped key name" used for comparing keys
- the "key name" which will be used to create keys/GUIs/CLIs/...

Maybe `keyGetEscapedName(Key *, char *, size_t len)?? and keyGetEscapedBaseName? Or which API do you propose?

Reply:
> Yes, but we definitely cannot call it store/load and the API is essential here for this decision (does it integrate nicely to the other API functions).

What is the problem with using the placeholders `store` and `load` for now? If you prefer we could use `storageEncode` and `storageDecode`.

> Do I understand correctly that in essence there would be 3 different forms of key names:

We would still only store the (canonical) escaped key name in `key->key` and the unescaped key name in `key->ukey`. They both serve the same purpose as before, but unescaped key names will be interpreted a bit differently.

Before the meaning of an unescaped key name essentially depended on the context. `#0` might be the first array element, if it is accessed or specified as such, but it might also just be a key part that is called `#0`; no other meaning at all.
Now every unescaped key name has a well-defined intended meaning. `#0` is intended to be the first array element. If it isn't used as such, this is could be considered a bug. The important idea here is that `\#0` is intended to be just a key part named `#0`. The same goes for any other unescaped part starting with `\` its intended meaning is just the rest of the string.

Storage plugins could also still use `key->key` for storage (like `quickdump` and `dump` do for example).

If they are more JSON-like and need to deconstruct the key name, simply splitting into unescaped parts also still works. But then there might be confusion around things that can be represented natively in the format.

For example in a format that has native arrays there is no good reason to interpret the name `#0` as the first array element, but if the plugin simply splits into unescaped parts, this would be the case.

> the "key name" which will be used to create keys/GUIs/CLIs/...

For creating the same (canonical or non-canonical) escaped key name format as ever is used.

How keys are displayed in user interfaces entirely depends on the interface.

A CLI or any other interface that represents a key in a single string should use the same format as `keyNew`, which has exactly these requirements.

A GUI can have so many shapes and forms, it is hard to say what should be used.
If the GUI is table-like the escaped name lends itself, as it is a single string that can be put into a single column.
If tree-like hierarchies are used splitting the unescaped name into parts and using these as node names is a good start. How these unescaped parts are shown is a similar question to how are field names in a JSON-like nested object format mapped to key parts. You could just use the exact unescaped parts, or you could create native representations for some or all special parts (array-elements in blue text, parts starting with dot on red background, etc). In which case you can remove leading backslashes just like you would in storage plugins.

> Maybe `keyGetEscapedName(Key *, char *, size_t len)?? and keyGetEscapedBaseName? Or which API do you propose?

Could we please first decide whether or not we even want to do this at all. It is pointless to discuss about function names, if we don't even know whether or not we are going to to do anything.